src/Corax/Mappings/IndexFieldsMappingBuilder.cs

```diff
@@ -0,0 +1,237 @@
+        }
+
+        return this;
+    }
+
+    // This should create deep copy
+    public IndexFieldsMappingBuilder AddBinding(IndexFieldBinding binding)
+    {
+        var fieldId = binding.FieldId;
+        var fieldName = binding.FieldName;
+
+        var canAdd = _fieldsById.ContainsKey(fieldId) == false && _fields.ContainsKey(fieldName) == false;
+
+        if (canAdd)
+        {
+            var bindingDeepCopy = binding.Copy(_context);
+            _fields[fieldName] = bindingDeepCopy;
+            _fieldsById[fieldId] = bindingDeepCopy;
+        }
+        else
+        {
+            throw new InvalidDataException("Cannot add already existing item into index mapping.");
+        }
+
+        return this;
+    }
+
+    private void GetFieldNameForLongs(Slice fieldName, out Slice fieldNameForLongs)
+    {
+        GetFieldNameWithPostfix(_context, fieldName, LongSuffix, out fieldNameForLongs);
+    }
+
+    internal static void GetFieldNameForLongs(ByteStringContext context, Slice fieldName, out Slice fieldNameForLongs)
+    {
+        GetFieldNameWithPostfix(context, fieldName, LongSuffix, out fieldNameForLongs);
+    }
+
+    
+    private void GetFieldNameForDoubles(Slice fieldName, out Slice fieldNameForDoubles)
+    {
+        GetFieldNameWithPostfix(_context, fieldName, DoubleSuffix, out fieldNameForDoubles);
+    }
+    
+    internal static void GetFieldNameForDoubles(ByteStringContext context, Slice fieldName, out Slice fieldNameForDoubles)
+    {
+        GetFieldNameWithPostfix(context, fieldName, DoubleSuffix, out fieldNameForDoubles);
+    }
+
+    private static unsafe void GetFieldNameWithPostfix(ByteStringContext context, Slice fieldName, short postfix, out Slice fieldWithPostfix)
+    {
+        context.Allocate(fieldName.Size + sizeof(short), out ByteString output);
+        fieldName.Content.CopyTo(output.Ptr);
+        *(short*)(output.Ptr + fieldName.Size) = postfix;
+        fieldWithPostfix = new Slice(SliceOptions.Key, output);
+    }
+
+    public IndexFieldsMappingBuilder AddSearchAnalyzer(Func<string, Analyzer> searchAnalyzer)
+    {
+        if (_searchAnalyzer != null)
+            throw new InvalidOperationException($"Search analyzer is already added");
+
+        _searchAnalyzer = searchAnalyzer;
+        return this;
+    }
+
+    public IndexFieldsMappingBuilder AddExactAnalyzer(Func<string, Analyzer> exactAnalyzer)
+    {
+        if (_exactAnalyzer != null)
+            throw new InvalidOperationException($"Exact analyzer is already added");
+
+        _exactAnalyzer = exactAnalyzer;
+        return this;
+    }
+
+    public IndexFieldsMappingBuilder AddDefaultAnalyzer(Analyzer defaultAnalyzer)
+    {
+        if (_exactAnalyzer != null)
+            throw new InvalidOperationException($"Exact analyzer is already added");
+
+        _defaultAnalyzer = defaultAnalyzer;
+        return this;
+    }
+    
+    public IndexFieldsMapping Build()
+    {
+        if (_materialized && _isDynamic == false)
```

Comment:
And if it _is_ dynamic? 
Don't we have multiple mapping instances with the same `_context` ?

Reply:
When it is dynamic we just update the bindings with the same context. Calling `Dispose` on the last updated Mapping will clear all allocations.

Reply:
So we aren't keeping the older instance around then, nor disposing it?

Reply:
Yes.
I've changed the condition to ` (_materialized && (_isForWriter && _isDynamic) == false)` to be more accurate.

The only way we allow to add items and materialize is when we're in a single thread (indexing) and have dynamic fields.

Reusing the allocator gives us the possibility to add only new things without disposing previous mapping (because it is in the same memory, there is no need to perform deep copy)